### PR TITLE
refactor predict for dnn model

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

1. refactor predictImage and predict to support dnn model inference
2. refactor evaluate to support dnn model inference
3. refactor local predict to support dnn model inference
4. add Imagenet test example
5. refactor IRGraph to support run dnn thread inside.

## How was this patch tested?
unit tests and use trained model to predict by hand

### CODE REVIEW

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/DnnPredictorSpec.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+      val ir = inception.cloneModule().toGraph().asInstanceOf[StaticGraph[Float]].toIRgraph()
+      val in = data(0).feature.reshape(Array(1, 3, 224, 224))
+
+      t.evaluate()
+      inception.evaluate()
+      ir.evaluate()
+
+      val out1 = t.forward(in.clone())
+      val out2 = inception.forward(in.clone())
+      val out3 = ir.forward(in.clone())
```
Comment (qiuxin2012) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/DnnPredictorSpec.scala]: what are out1 2 3 doing?

Comment (zhangxiaoli73) [reply spark/dl/src/test/scala/com/intel/analytics/bigdl/utils/intermediate/DnnPredictorSpec.scala]: test code, deleted

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TestImageNet.scala [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import com.intel.analytics.bigdl.optim._
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric._
+import com.intel.analytics.bigdl.transform.vision.image.{ImageFeature, MTImageFeatureToBatch, MatToTensor, PixelBytesToMat}
+import com.intel.analytics.bigdl.transform.vision.image.augmentation.{ChannelScaledNormalizer, RandomCropper, RandomResize}
+import com.intel.analytics.bigdl.utils._
+import org.apache.log4j.{Level, Logger}
+import org.apache.spark.SparkContext
+
+object TestImageNet {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TestImageNet.scala]: add comments for the example?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala [hunk 3]
```diff
@@ -115,6 +122,7 @@ private[bigdl] class ModelBroadcastImp[T: ClassTag](applyProtoBuffer: Boolean =
       putWeightBias(SerializationUtils.clone(weightsBias), model)
       initGradWeightBias(weightsBias, model)
     }
+    this.setNodeAndCore()
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: remove `this` and `()` ? to keep common style

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Evaluator.scala [hunk 4]
```diff
@@ -71,4 +76,34 @@ class Evaluator[T: ClassTag] private[optim](model: Module[T])(implicit ev: Tenso
         left.zip(right).map { case (l, r) => l + r }
     }).zip(vMethods)
   }
+
+  /**
+   * Applies ValidationMethod to the model and rdd dataset.
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Evaluator.scala]: Applies -> Apply

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala [hunk 5]
```diff
@@ -55,28 +56,44 @@ private[bigdl] class IRGraph[T: ClassTag](
 
   private var graph: Graph[T] = null
 
+  private[bigdl] def isBuild(): Boolean = if (graph == null) false else true
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: change to 

`private[bigdl] def isBuild(): Boolean = graph != null`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala (hunk 3 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/utils/ModelBroadcast.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Evaluator.scala (hunk 4 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/Evaluator.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/intermediate/IRGraph.scala]: ok, changed

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TestImageNet.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/models/resnet/TestImageNet.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/379

**Comment by zhangxiaoli73** [comment]
jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/view/PR/job/BigDL-PRVN-lite/388/

**Comment by wzhongyuan** [comment]
LGTM

**Pull request merged**